File: Source/Engine/HyperCore/src/HyperCore/Logger.cpp
/*
 * Copyright (c) 2020-2021, SkillerRaptor <skillerraptor@protonmail.com>
 *
 * SPDX-License-Identifier: MIT
 */

#include "HyperCore/Logger.hpp"

#include "HyperCore/Prerequisites.hpp"

#include <fmt/chrono.h>
#include <fmt/color.h>

namespace HyperEngine
{
#if HYPERENGINE_DEBUG
	Logger::Level Logger::s_log_level = Logger::Level::Trace;
#else
	Logger::Level Logger::s_log_level = Logger::Level::Fatal;
#endif

	auto Logger::internal_log(const Level level, const std::string_view string) -> void
	{
		const auto level_color = [&level]() -> fmt::color
		{
			switch (level)
			{
			case Level::Info:
				return fmt::color::white;
			case Level::Warning:
				return fmt::color::gold;
			case Level::Error:
				return fmt::color::red;
			case Level::Fatal:
				return fmt::color::crimson;
			case Level::Debug:
				return fmt::color::gray;
			default:
				return fmt::color::white;
			}
		}();

		const auto level_name = [&level]() -> const char*
		{
			switch (level)
			{
			case Level::Info:
				return "Info";
			case Level::Warning:
				return "Warning";
			case Level::Error:
				return "Error";
			case Level::Fatal:
				return "Fatal";
			case Level::Debug:
				return "Debug";
			default:
				return "Undefined";
			}
		}();

		fmt::print(
			"{} | {}: {}\n",
			fmt::format(fmt::fg(fmt::color::dark_gray), "{:%H:%M:%S}", fmt::localtime(std::time(nullptr))),
			fmt::format(fmt::fg(level_color), level_name),
			string);
	}
	
	auto Logger::set_level(const Logger::Level log_level) -> void
	{
		s_log_level = log_level;
	}
	
	auto Logger::level() -> Logger::Level
	{
		return s_log_level;
	}
} // namespace HyperEngine

File: crates/hyper_editor/src/logger.rs
/*
 * Copyright (c) 2022-2024, SkillerRaptor
 *
 * SPDX-License-Identifier: MIT
*/

use std::io;

use anyhow::Result;
use chrono::Local;
use fern::{
    colors::{Color, ColoredLevelConfig},
    Dispatch,
};
use log::{Level, LevelFilter};

#[allow(dead_code)]
#[derive(Clone, Copy, Debug)]
pub(crate) enum Verbosity {
    Error,
    Warning,
    Info,
    Debug,
    Trace,
}

pub(crate) fn init(verbosity: Verbosity) -> Result<()> {
    let levels = ColoredLevelConfig::new()
        .error(Color::Red)
        .warn(Color::BrightYellow)
        .info(Color::Green)
        .debug(Color::Cyan)
        .trace(Color::Magenta);

    let level_filter = match verbosity {
        Verbosity::Error => LevelFilter::Error,
        Verbosity::Warning => LevelFilter::Warn,
        Verbosity::Info => LevelFilter::Info,
        Verbosity::Debug => LevelFilter::Debug,
        Verbosity::Trace => LevelFilter::Trace,
    };

    Dispatch::new()
        .format(move |out, message, record| {
            let reset = "\x1B[0m";

            let time = {
                let current_time = Local::now().format("%Y-%m-%d %H:%M:%S");
                let bright_black = format!("\x1B[{}m", Color::BrightBlack.to_fg_str());
                format!("{}[{}]{}", bright_black, current_time, reset)
            };

            let target = {
                let current_target = record.target().split_once(':');
                let white = format!("\x1B[{}m", Color::White.to_fg_str());
                format!("{}{:<15}{}", white, current_target.unwrap().0, reset)
            };

            let level = {
                let current_level = record.level();
                let color = format!("\x1B[{}m", levels.get_color(&record.level()).to_fg_str());
                format!("{}{:<5}{}", color, current_level, reset)
            };

            let message = {
                let color = if record.level() == Level::Trace {
                    format!("\x1B[{}m", Color::White.to_fg_str())
                } else {
                    format!("\x1B[{}m", Color::BrightWhite.to_fg_str())
                };

                format!("{}{}", color, message)
            };

            out.finish(format_args!(
                "{} {} {} {} {}",
                time, target, level, message, reset
            ))
        })
        .level(LevelFilter::Off)
        .level_for("hyper_core", level_filter)
        .level_for("hyper_ecs", level_filter)
        .level_for("hyper_editor", level_filter)
        .level_for("hyper_engine", level_filter)
        .level_for("hyper_game", level_filter)
        .level_for("hyper_math", level_filter)
        .level_for("hyper_platform", level_filter)
        .level_for("hyper_render", level_filter)
        .level_for("hyper_vulkan", level_filter)
        .chain(io::stdout())
        .apply()?;

    Ok(())
}
